GeoLib/SurfaceGrid.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * \date 2012-09-22
+ * \brief Definition of the SurfaceGrid class.
+ *
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ */
+
+#include <algorithm>
+#include <bitset>
+
+#include <logog/include/logog.hpp>
+
+#include "SurfaceGrid.h"
+
+#include "Surface.h"
+#include "Triangle.h"
+
+#include "MathLib/Point3d.h"
+
+namespace GeoLib {
+
+SurfaceGrid::SurfaceGrid(Surface const*const sfc) :
+	AABB(sfc->getAABB()), _n_steps({{1,1,1}})
+{
+	// enlarge the bounding, such that the points with maximal coordinates
+	// fits into the grid
+	for (std::size_t k(0); k<3; ++k) {
+		_max_pnt[k] += std::abs(_max_pnt[k]) * 1e-6;
+		if (std::abs(_max_pnt[k]) < std::numeric_limits<double>::epsilon()) {
+			_max_pnt[k] = (_max_pnt[k] - _min_pnt[k]) * (1.0 + 1e-6);
+		}
+	}
+
+	std::array<double, 3> delta{{ _max_pnt[0] - _min_pnt[0],
+		_max_pnt[1] - _min_pnt[1], _max_pnt[2] - _min_pnt[2] }};
+
+	if (delta[0] < std::numeric_limits<double>::epsilon()) {
+		const double max_delta(std::max(delta[1], delta[2]));
+		_min_pnt[0] -= max_delta * 0.5e-3;
+		_max_pnt[0] += max_delta * 0.5e-3;
+		delta[0] = _max_pnt[0] - _min_pnt[0];
+	}
+
+	if (delta[1] < std::numeric_limits<double>::epsilon()) {
+		const double max_delta(std::max(delta[0], delta[2]));
+		_min_pnt[1] -= max_delta * 0.5e-3;
+		_max_pnt[1] += max_delta * 0.5e-3;
+		delta[1] = _max_pnt[1] - _min_pnt[1];
+	}
+
+	if (delta[2] < std::numeric_limits<double>::epsilon()) {
+		const double max_delta(std::max(delta[0], delta[1]));
+		_min_pnt[2] -= max_delta * 0.5e-3;
+		_max_pnt[2] += max_delta * 0.5e-3;
+		delta[2] = _max_pnt[2] - _min_pnt[2];
+	}
+
+	const std::size_t n_tris(sfc->getNTriangles());
+	const std::size_t n_tris_per_cell(5);
+
+	std::bitset<3> dim; // all bits are set to zero.
+	for (std::size_t k(0); k<3; ++k)
+		if (std::abs(delta[k]) >= std::numeric_limits<double>::epsilon())
+			dim[k] = true;
+
+	// *** condition: n_tris / n_cells < n_tris_per_cell
+	//                where n_cells = _n_steps[0] * _n_steps[1] * _n_steps[2]
+	// *** with _n_steps[0] = ceil(pow(n_tris*delta[0]*delta[0]/(n_tris_per_cell*delta[1]*delta[2]), 1/3.)));
+	//          _n_steps[1] = _n_steps[0] * delta[1]/delta[0],
+	//          _n_steps[2] = _n_steps[0] * delta[2]/delta[0]
+	auto ceil = [](double v){
+		return static_cast<std::size_t>(std::ceil(v));
+	};
+	switch (dim.count()) {
+	case 3: // 3d case
+		_n_steps[0] = ceil(std::cbrt(
```

Comment:
This can be done later in a separate PR. `sc_ceil` will be corrected in some minutes.